utilities/lookup.py

```diff
@@ -0,0 +1,77 @@
+from typing import Union
+
+
+class LookupTable:
+    def __init__(self, **kwargs: Union[list[float], tuple[float]]) -> None:
+        if len(kwargs.keys()) < 2:
+            raise ValueError("LookupTable: Must have more than one row!")
+
+        key_found = False
+        self.rows = {}
+        for key, val in kwargs.items():
+            # Find length of table based on first row, which should be the key
+            if not key_found:
+                self.key_name = key
+
+                if val != sorted(val):
+                    raise ValueError(
+                        "LookupTable: Key values must be in ascending order!"
+                    )
+
+                self.key_values = tuple(val)
+                self.table_length = len(val)
+
+                if self.table_length < 2:
+                    raise ValueError("LookupTable: Should have more than one value!")
+
+                key_found = True
+                continue
+
+            # Convert to tuple so values are immutable
+            self.rows[key] = tuple(val)
+
+            if len(val) != self.table_length:
+                raise IndexError(
+                    f"LookupTable: Length of row '{key}' does not match length of key row!"
+                )
+
+    def lookup(
+        self, lookup_row: str, lookup_value: float, extrapolate: bool = False
+    ) -> float:
+        # Row name not in table
+        if lookup_row not in self.rows:
+            return 0.0
+
+        # Go through the key values until we pass value
+        start_index = 0
+        for i, value in enumerate(self.key_values):
+            if value > lookup_value:
+                break
+            start_index = i
+
+        if extrapolate:
```

Comment:
The variable name is misleading, we still extrapolate even if this is false, we just use a constant rather than linear scheme to do so. I think this is an option we want to have. I don't know if we would use a true no extrapolation mode as we would have to return bad values or throw errors.